File: FList.cpp
#include <iostream>
#include <string>
#include "FList.h"

using namespace std;

void FCREATE(FList& new_FList, const string& data){
    new_FList.head = new FLNode(data, nullptr);
}

FLNode* FGETINDEX(FList& list, int index){
    FLNode* Address = list.head;
    int currentIndex = 0;
    
    while (Address != nullptr) {
        if (currentIndex == index) {
            return Address;
        }
        Address = Address->next;
        currentIndex++;
    }
    
    cout << "Элемента с таким индексом нет" << endl;
    exit(1);
}

FLNode* FGETFLNodePARENT(FList& list, FLNode* target_FLNode){
    if (list.head == nullptr || list.head == target_FLNode) {
        return list.head;
    }
    
    FLNode* Address = list.head;
    while (Address->next != nullptr) {
        if (Address->next == target_FLNode) {
            return Address;
        }
        Address = Address->next;
    }
    
    cout << "Узел отсутствует в списке" << endl;
    exit(1);
}

FLNode* FGETDATA(FList& list, const string& data){
    FLNode* Address = list.head;
    while (Address != nullptr) {
        if (Address->data == data) {
            return Address;
        }
        Address = Address->next;
    }
    
    cout << "Узел отсутствует в списке" << endl;
    exit(1);
}

void FADDTAIL(FList& list, const string& data){
    FLNode* new_FLNode = new FLNode(data, nullptr);
    
    if (list.head == nullptr) {
        list.head = new_FLNode;
        return;
    }
    
    FLNode* Address = list.head;
    while (Address->next != nullptr) {
        Address = Address->next;
    }
    Address->next = new_FLNode;
}

void FADDHEAD(FList& list, const string& data){
    FLNode* new_FLNode = new FLNode(data, list.head);
    list.head = new_FLNode;
}

void FADDAFTER(FLNode* ptr, const string& data){
    FLNode* new_FLNode = new FLNode(data, ptr->next);
    ptr->next = new_FLNode;
}

void FADDBEFORE(FList& list, FLNode* ptr, const string& data){
    if (list.head == ptr){
        FADDHEAD(list, data);
        return;
    }
    
    FLNode* parent = FGETFLNodePARENT(list, ptr);
    FLNode* new_FLNode = new FLNode(data, ptr);
    parent->next = new_FLNode;
}

void FDELTAIL(FList& list){
    if (list.head == nullptr) return;
    
    if (list.head->next == nullptr) {
        delete list.head;
        list.head = nullptr;
        return;
    }
    
    FLNode* Address = list.head;
    while (Address->next->next != nullptr) {
        Address = Address->next;
    }
    
    delete Address->next;
    Address->next = nullptr;
}

void FDELHEAD(FList& list){
    if (list.head == nullptr) return;
    
    FLNode* Address = list.head->next;
    delete list.head;
    list.head = Address;
}

void FDELAFTER(FLNode* ptr){
    if (ptr->next == nullptr) return;
    
    FLNode* Address = ptr->next->next;
    delete ptr->next;
    ptr->next = Address;
}

void FDELBEFORE(FList& list, FLNode* ptr){
    if (list.head == ptr){
        cout << "Элемент первый в списке" << endl;
        return;
    }
    
    if (list.head->next == ptr){
        FDELHEAD(list);
        return;
    }
    
    FLNode* grandParent = FGETFLNodePARENT(list, FGETFLNodePARENT(list, ptr));
    FLNode* parent = grandParent->next;
    grandParent->next = ptr;
    delete parent;
}

void FDELDATA(FList& list, const string& data){
    if (list.head == nullptr) {
        cout << "Список пуст" << endl;
        return;
    }
    
    if (list.head->data == data){
        FDELHEAD(list);
        return;
    }

    FLNode* Address = list.head;
    while (Address->next != nullptr) {
        if (Address->next->data == data) {
            FDELAFTER(Address);
            return;
        }
        Address = Address->next;
    }
    
    cout << "Элемента нет в списке" << endl;
}

bool isListValid(FList& list) {
    if (list.head == nullptr) return true;
    
    FLNode* slow = list.head;
    FLNode* fast = list.head;
    
    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
        
        if (slow == fast) {
            cout << "DEBUG: Cycle detected in list!" << endl;
            return false;
        }
    }
    return true;
}

void FPRINT(FList& list){
    FLNode* Address = list.head;
    while (1){
        if(Address -> next == nullptr){
            cout << Address -> data << endl;
            return;
        }
        else{
            cout << Address -> data << endl;
            Address = Address -> next;
        }
    }
}

File: FList.go
package main

import (
	"fmt"
)

type FLNode struct {
	data string
	next *FLNode
}

// Конструктор
func NewFLNode(val string, nextPtr *FLNode) *FLNode {
	return &FLNode{
		data: val,
		next: nextPtr,
	}
}

type FList struct {
	head *FLNode
}

// Конструктор
func NewFList() *FList {
	return &FList{
		head: nil,
	}
}

// Конструктор копирования
func (f *FList) Copy() *FList {
	newList := &FList{
		head: nil,
	}

	current := f.head
	var last *FLNode

	for current != nil {
		newNode := NewFLNode(current.data, nil)
		if newList.head == nil {
			newList.head = newNode
		} else {
			last.next = newNode
		}
		last = newNode
		current = current.next
	}

	return newList
}

// Аналог оператора присваивания
func (f *FList) Assign(other *FList) *FList {
	if f != other {
		// Копируем элементы
		current := other.head
		var last *FLNode

		for current != nil {
			newNode := NewFLNode(current.data, nil)
			if f.head == nil {
				f.head = newNode
			} else {
				last.next = newNode
			}
			last = newNode
			current = current.next
		}
	}
	return f
}

func FCREATE(newFList *FList, data string) {
	newFList.head = NewFLNode(data, nil)
}

func FGETINDEX(list *FList, index int) *FLNode {
	Address := list.head
	currentIndex := 0

	for Address != nil {
		if currentIndex == index {
			return Address
		}
		Address = Address.next
		currentIndex++
	}

	fmt.Println("Элемента с таким индексом нет")
	return nil
}

func FGETFLNodePARENT(list *FList, targetFLNode *FLNode) *FLNode {
	if list.head == nil || list.head == targetFLNode {
		return list.head
	}

	Address := list.head
	for Address.next != nil {
		if Address.next == targetFLNode {
			return Address
		}
		Address = Address.next
	}

	fmt.Println("Узел отсутствует в списке")
	return nil
}

func FGETDATA(list *FList, data string) *FLNode {
	Address := list.head
	for Address != nil {
		if Address.data == data {
			return Address
		}
		Address = Address.next
	}

	fmt.Println("Узел отсутствует в списке")
	return nil
}

func FADDTAIL(list *FList, data string) {
	newFLNode := NewFLNode(data, nil)

	if list.head == nil {
		list.head = newFLNode
		return
	}

	Address := list.head
	for Address.next != nil {
		Address = Address.next
	}
	Address.next = newFLNode
}

func FADDHEAD(list *FList, data string) {
	newFLNode := NewFLNode(data, list.head)
	list.head = newFLNode
}

func FADDAFTER(ptr *FLNode, data string) {
	newFLNode := NewFLNode(data, ptr.next)
	ptr.next = newFLNode
}

func FADDBEFORE(list *FList, ptr *FLNode, data string) {
	if list.head == ptr {
		FADDHEAD(list, data)
		return
	}

	parent := FGETFLNodePARENT(list, ptr)
	if parent != nil {
		newFLNode := NewFLNode(data, ptr)
		parent.next = newFLNode
	}
}

func FDELTAIL(list *FList) {
	if list.head == nil {
		return
	}

	if list.head.next == nil {
		list.head = nil
		return
	}

	Address := list.head
	for Address.next.next != nil {
		Address = Address.next
	}

	Address.next = nil
}

func FDELHEAD(list *FList) {
	if list.head == nil {
		return
	}

	Address := list.head.next
	list.head = Address
}

func FDELAFTER(ptr *FLNode) {
	if ptr.next == nil {
		return
	}

	Address := ptr.next.next
	ptr.next = Address
}

func FDELBEFORE(list *FList, ptr *FLNode) {
	if list.head == ptr {
		fmt.Println("Элемент первый в списке")
		return
	}

	if list.head.next == ptr {
		FDELHEAD(list)
		return
	}

	parent := FGETFLNodePARENT(list, ptr)
	if parent != nil {
		grandParent := FGETFLNodePARENT(list, parent)
		if grandParent != nil {
			grandParent.next = ptr
		}
	}
}

func FDELDATA(list *FList, data string) {
	if list.head == nil {
		fmt.Println("Список пуст")
		return
	}

	if list.head.data == data {
		FDELHEAD(list)
		return
	}

	Address := list.head
	for Address.next != nil {
		if Address.next.data == data {
			FDELAFTER(Address)
			return
		}
		Address = Address.next
	}

	fmt.Println("Элемента нет в списке")
}

func isListValid(list *FList) bool {
	if list.head == nil {
		return true
	}

	slow := list.head
	fast := list.head

	for fast != nil && fast.next != nil {
		slow = slow.next
		fast = fast.next.next

		if slow == fast {
			fmt.Println("DEBUG: Cycle detected in list!")
			return false
		}
	}
	return true
}

func FPRINT(list *FList) {
	Address := list.head
	for {
		if Address.next == nil {
			fmt.Println(Address.data)
			return
		} else {
			fmt.Println(Address.data)
			Address = Address.next
		}
	}
}
